src/v/security/audit/schemas/hashing_utils.cc

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "security/audit/schemas/hashing_utils.h"
+
+#include "hashing/utils.h"
+#include "security/acl.h"
+#include "security/audit/schemas/application_activity.h"
+#include "security/audit/schemas/iam.h"
+
+namespace security::audit {
+
+template<typename T>
+size_t audit_type_base_hash(
+  category_uid ct_uid, class_uid c_uid, severity_id s_id, T activity_id) {
+    /// metadata field explicity omitted since it is static
+    return hash_combine(ct_uid, c_uid, s_id, get_ocsf_type(c_uid, activity_id));
+}
+
+size_t actor_hash(const security::auth_result& result) {
+    /// user
+    size_t h = hash_combine(result.principal.name(), result.is_superuser);
+
+    // policy
+    enum class auth_state { disabled = 0, superuser = 1, nomatches = 2 };
+    using enum_t = std::underlying_type_t<auth_state>;
+    if (result.authorization_disabled) {
+        hash_combine(h, static_cast<enum_t>(auth_state::disabled));
+    } else if (result.is_superuser) {
+        hash_combine(h, static_cast<enum_t>(auth_state::superuser));
+    } else if (result.empty_matches) {
+        hash_combine(h, static_cast<enum_t>(auth_state::nomatches));
+    } else {
+        hash_combine(h, result.acl);
+        hash_combine(h, result.resource_pattern);
+    }
+    /// authorization result (decision + policy)
+    return hash_combine(h, result.authorized);
+}
+
+size_t api_activity_event_base_hash(
+  std::string_view operation_name,
+  const security::auth_result& auth_result,
+  const ss::socket_address& local_address,
+  std::string_view service_name,
+  ss::net::inet_address client_addr) {
+    auto crud = api_activity::op_to_crud(auth_result.operation);
+    size_t h = audit_type_base_hash(
+      category_uid::application_activity,
+      class_uid::api_activity,
+      severity_id::informational,
+      crud);
+    return hash_combine(
+      h,
+      actor_hash(auth_result),
+      operation_name,
+      service_name,
+      local_address.addr(),
+      client_addr,
+      auth_result.authorized,
+      auth_result.acl,
+      auth_result.resource_pattern);
+}
+
+size_t http_request_hash(const ss::http::request& req) {
+    size_t h = 0;
+    /// request headers
+    for (const auto& hdr : req._headers) {
+        hash_combine(h, hdr.first);
+        hash_combine(h, hdr.second);
+    }
```

Comment:
Is there a concern if the headers appear in a different order? I.e., should they be sorted or otherwise normalised?

Also:
```suggestion
    hash_combine(h, req._headers);
```

Reply:
no order doesn't matter